### ISSUE TITLE
Conan ignores (os and compiler) settings while resolving version ranges

### ISSUE THREAD

**Issue by Aalmann** [conan-io/conan#3470]
Hi @lasote and @memsharded,

I found a strange behavior and hope you can help.

### Given
I have a package A which depends on a version range of package B.
Package B is available in different versions and for different compiler versions, but not all combinations are available (e.g. due to deprecated compiler versions). 
Availability matrix for package B (published to artifactory):
- Version 1.0.0: 
  * Windows vc90 
  * Windows vc100 
  * Windows vc140 
  * Linux gcc 4.6 
  * Linux gcc 4.8 
  * Linux gcc 5.4
- Version 1.0.1: 
  * Windows vc90 
  * Windows vc100 
  * Windows vc140 
  * Linux gcc 4.6 
  * Linux gcc 4.8 
  * Linux gcc 5.4
- Version 1.1.0: 
  * ~~Windows vc90~~ (compiler version no longer available)
  * Windows vc100 
  * Windows vc140 
  * Linux gcc 4.6 
  * Linux gcc 4.8 
  * Linux gcc 5.4

Package A contains `self.requires("package_B/[>=1.0.0,<2.0.0]@user/channel")` as version range definition.

### My expectation
For all os-compiler-combinations (without Windows vc90) version 1.1.0 will be resolved. Make sense to me, because 1.1.0 is latest and greatest for the most of the packages.
For Windows vc90 version 1.0.1 will be resolved, because it is the latest and greatest for this os-compiler-combination.

### Problem
But the error message of conan is something like "Missing prebuilt package 'package_B/1.1.0@user/channel' ...", so it seems to be that conan ignores the os and compiler settings while getting the latest and greatest available version of a package.
IMHO conan has to check which versions are available (get list of versions), then check if a package for the given settings is available and if not, check the next lower version.

### conan versions tested
* 1.5.2
* 1.7.1

Thanks,
Best Aalmann

**Comment by memsharded** [comment]
Hi @Aalmann 

This is almost impossible due to the combinatoric explosion. The version ranges are resolved, the same as the whole dependency graph, irrespective of the binaries that could exist or not. Then, in a later stage, conan checks for binary availability. This makes dependency resolution linear in the size of the graph. 

If binaries availability has to be checked, it is an NP problem (known as SAT solving), in which for every recipe, the existing binaries for every version in the range has to be requested to the server (a lot of slow API calls), and that explodes as you are building the dependency graph upstream. That compounds with the fact that to know which binary do you need (the package-ID), you need to know first the whole dependency graph, as the package-ID is a function not only of the settings and options, but also of the dependencies, could include the options and settings of the dependencies too. That makes the problem intractable in practical terms.

I would say the only viable/practical solution would be explicitly stating deps in:

```python
def requirements(self):
    if self.settings.compiler.version > "9.0":
        self.requires("package_B/[>=1.0.0,<2.0.0]@user/channel")
   else:
       self.requires("package_B/[>=1.0.0,<1.1.0]@user/channel")
```

**Comment by memsharded** [comment]
> Problem with this is that package A has to know something about the future of package B.

Yes, but that is not a problem, that is a definition intrinsic to version ranges. By definition, if you are using version ranges, you are telling to the package manager: "I can use any of the versions that are in this range". If the upstream package B breaks the compatibility contract defined by its version it should certainly bump version to something that is outside of that range. Otherwise, package B not providing a certain binary for a version that is in the version range is a violation of the contract, and the expected is an error, not trying to fallback to a previous existing version that indeed provides such a binary. 

And even if we wanted to implement such approach, there is still the NP-hard issue. I don't think that complexity can be solved in reasonable times. So it is not only that it doesn't sound the best approach, but it is also not-affordable in computational terms.
